### PR TITLE
Account for Eth block re-orgs in eth indexer

### PR DESCRIPTION
**Pull request by Bridgerz**
Description
Account for Eth block re-orgs in eth indexer #19158

Test plan
Spot check on my local machine. Indexer validity checker tool coming soon...

### CODE REVIEW

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs [hunk 1]
```diff
@@ -183,6 +183,25 @@ impl Datasource<RawEthData> for EthSyncDatasource {
         let handle = spawn_monitored_task!(async move {
             let mut cached_blocks: HashMap<u64, Block<H256>> = HashMap::new();
 
+            let mut last_finalized_block = retry_with_max_elapsed_time!(
+                client.get_last_finalized_block_id(),
+                Duration::from_secs(30000)
+            )
+            .expect("Unable to get latest finalized block from provider")
+            .expect("Failed to unwrap the inner result");
```
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:191]: rather than having multiple tasks query this, we should have one dedicated thread check this and update the watch channel, just like what we do for EthSyncer

**Comment by longbowlu** [comment]
now we are a "live" finalized task, we do want a `live_task`, right?

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs [hunk 2]
```diff
@@ -33,20 +35,23 @@ use crate::{
...
-        eth_sui_bridge_contract_address: String,
+        eth_sui_bridge_contract_addresses: Vec<String>,
         eth_ws_url: String,
         indexer_metrics: BridgeIndexerMetrics,
     ) -> Result<Self, anyhow::Error> {
-        let bridge_address = EthAddress::from_str(&eth_sui_bridge_contract_address)?;
+        let bridge_addresses = eth_sui_bridge_contract_addresses
+            .iter()
+            .map(|address| EthAddress::from_str(address).unwrap())
+            .collect();
```
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:52]: unwrap here is unsafe. Do sth like this
```
        let bridge_addresses = eth_sui_bridge_contract_addresses
            .iter()
            .map(|address| EthAddress::from_str(address))
            .collect::<Result<_>>()?;
```
which returns the error as long as there is at least one error

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs [hunk 3]
```diff
@@ -131,22 +136,25 @@ impl Datasource<RawEthData> for EthSubscriptionDatasource {
...
-        eth_sui_bridge_contract_address: String,
+        eth_sui_bridge_contract_addresses: Vec<String>,
         eth_http_url: String,
         indexer_metrics: BridgeIndexerMetrics,
         bridge_metrics: Arc<BridgeMetrics>,
     ) -> Result<Self, anyhow::Error> {
-        let bridge_address = EthAddress::from_str(&eth_sui_bridge_contract_address)?;
+        let bridge_addresses = eth_sui_bridge_contract_addresses
+            .iter()
+            .map(|address| EthAddress::from_str(address).unwrap())
```
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:154]: ditto

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs [hunk 4]
```diff
@@ -175,87 +183,151 @@ impl Datasource<RawEthData> for EthFinalizedSyncDatasource {
                 .interval(std::time::Duration::from_millis(2000)),
         );
 
-        let bridge_address = self.bridge_address;
+        let bridge_addresses = self.bridge_addresses.clone();
         let indexer_metrics: BridgeIndexerMetrics = self.indexer_metrics.clone();
         let client = Arc::clone(&client);
         let provider = Arc::clone(&provider);
+        let bridge_metrics = Arc::clone(&self.bridge_metrics);
```
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:190]:
```
        let bridge_metrics = self.bridge_metrics.clone();
```

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs [hunk 5]
```diff
@@ -175,87 +183,151 @@ impl Datasource<RawEthData> for EthFinalizedSyncDatasource {
...
-            .expect("Failed to unwrap the inner result");
-
-            // block process until current range is finalized
-            while target_checkpoint > last_finalized_block {
-                tokio::time::sleep(Duration::from_secs(5)).await;
-
-                last_finalized_block = retry_with_max_elapsed_time!(
-                    client.get_last_finalized_block_id(),
-                    Duration::from_secs(30000)
+            if target_checkpoint > current_block {
```
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:194]: this feel a bit unsafe. and on sui side there are a few places like this which i don't quite like. Can we add a TODO here and we will come back with a consistent and robust alternative

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs [hunk 6]
```diff
@@ -175,87 +183,151 @@ impl Datasource<RawEthData> for EthFinalizedSyncDatasource {
...
+    client: Arc<EthClient<MeteredEthHttpProvier>>,
+    provider: Arc<Provider<Http>>,
+    addresses: Vec<EthAddress>,
+    starting_checkpoint: u64,
+    target_checkpoint: u64,
+    data_sender: DataSender<RawEthData>,
+    indexer_metrics: BridgeIndexerMetrics,
+) -> Result<(), Error> {
+    let Ok(Ok(logs)) = retry_with_max_elapsed_time!(
+        client.get_raw_events_in_range(addresses.clone(), starting_checkpoint, target_checkpoint),
```
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:274]: `fn get_raw_events_in_ranges` has this comment:
```
    // Note: query may fail if range is too big. Callsite is responsible
    // for chunking the query.
```

We need to segment it (e.g. 1k blocks per chunk) before calling `get_raw_events_in_range`

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs [hunk 7]
```diff
@@ -175,87 +183,151 @@ impl Datasource<RawEthData> for EthFinalizedSyncDatasource {
...
-                    panic!("Unable to get transaction from provider");
-                };
+async fn process_logs(
+    logs: Vec<RawEthLog>,
+    provider: Arc<Provider<Http>>,
+    data_sender: DataSender<RawEthData>,
+    target_checkpoint: u64,
+) -> Result<(), Error> {
+    let mut data = Vec::new();
+    let mut cached_blocks: HashMap<u64, Block<H256>> = HashMap::new();
```
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:302]: we have similar things in `EthSubscriptionDatasource`. I think we should have a shared cache between the subscription and the live sync task (for finalized events). let's add a TODO here

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs [hunk 8]
```diff
@@ -175,87 +183,151 @@ impl Datasource<RawEthData> for EthFinalizedSyncDatasource {
...
+    for log in logs {
```
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:303]: the way we do query `get_block` one by one here is not efficient. Let's add a TODO to add some parallelism

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs [hunk 9]
```diff
@@ -175,87 +183,151 @@ impl Datasource<RawEthData> for EthFinalizedSyncDatasource {
...
+                block_number: block,  // Handle block number conversion
+                tx_hash: log.tx_hash, // Assuming `event` contains `tx_hash`
+                log: log.log,
+            })
+            .collect();
+
+        process_logs(raw_logs, provider.clone(), data_sender.clone(), block)
+            .await
+            .expect("Failed to process logs");
+        indexer_metrics.latest_committed_eth_block.set(block as i64);
```
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:258]: are we using this for EthSubscriptionDatasource too?

File: crates/sui-bridge/src/eth_client.rs [hunk 10]
```diff
@@ -200,8 +200,8 @@ where
         // Safeguard check that all events are emitted from requested contract address
```
Comment (longbowlu) [inline crates/sui-bridge/src/eth_client.rs:200]: `contract addresses`

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs (hunk 9 above)
Comment (Bridgerz) [reply crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:258]: Forgot to update the metrics. will update.

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs [hunk 11]
```diff
@@ -294,6 +267,127 @@ impl Datasource<RawEthData> for EthSyncDatasource {
...
+        .await
+        .expect("Failed to start eth syncer");
+
+    // forward received events to the data sender
+    while let Some((_, block, logs)) = eth_events_rx.recv().await {
+        let raw_logs: Vec<RawEthLog> = logs
+            .into_iter()
+            .map(|log| RawEthLog {
+                block_number: block,  // Handle block number conversion
+                tx_hash: log.tx_hash, // Assuming `event` contains `tx_hash`
```
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:295]: these two lines of comments seem irrelevant or off.

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs [hunk 12]
```diff
@@ -294,6 +267,127 @@ impl Datasource<RawEthData> for EthSyncDatasource {
...
+                log: log.log,
+            })
+            .collect();
+
+        process_logs(raw_logs, provider.clone(), data_sender.clone(), block)
+            .await
+            .expect("Failed to process logs");
```
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:302]: do we really want to panic here? (we didn't panic in `retrieve_and_process_log_range`)

File: crates/sui-bridge-indexer/src/metrics.rs [hunk 13]
```diff
@@ -14,13 +14,11 @@ pub struct BridgeIndexerMetrics {
     pub(crate) total_sui_token_transfer_approved: IntCounter,
     pub(crate) total_sui_token_transfer_claimed: IntCounter,
     pub(crate) total_sui_bridge_txn_other: IntCounter,
+    pub(crate) last_committed_sui_checkpoint: IntGauge,
```
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/metrics.rs:17]: ditto for blame

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs (hunk 12 above)
Comment (Bridgerz) [reply crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:302]: I did this because we aren't returning a result from the live task, like the sync task does.

File: crates/sui-bridge-indexer/src/lib.rs [hunk 14]
```diff
@@ -49,6 +49,7 @@ pub struct TokenTransfer {
     gas_usage: i64,
     data_source: BridgeDataSource,
     data: Option<TokenTransferData>,
```
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/lib.rs:51]: could you remind me why this is optional?

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs [hunk 15]
```diff
@@ -202,63 +193,34 @@ impl Datasource<RawEthData> for EthSyncDatasource {
...
         let provider = provider.clone();
+        let bridge_metrics = self.bridge_metrics.clone();
 
         let handle = spawn_monitored_task!(async move {
-            let mut cached_blocks: HashMap<u64, Block<H256>> = HashMap::new();
-
-            let Ok(Ok(logs)) = retry_with_max_elapsed_time!(
-                client.get_raw_events_in_range(
-                    bridge_address,
+            // TODO: need a better way to determine if this task should be live or historical
```
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:202]: we can remove this TODO now

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs (hunk 12 above)
Comment (longbowlu) [reply crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:288]: not sure i follow. `retrieve_and_process_log_range` (is this what you meant by "sync task"?) does not return anything. One of them is `process_logs().await.expect()` and the other one is `process_logs().await?`

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs [hunk 16]
```diff
@@ -285,15 +247,155 @@ impl Datasource<RawEthData> for EthSyncDatasource {
...
+    is_finalized: bool,
+) -> Result<(), Error> {
+    let mut data = Vec::new();
+    let mut cached_blocks: HashMap<u64, Block<H256>> = HashMap::new();
+
+    println!(
+        "Processing {} logs, finalized: {}",
+        logs.len(),
+        is_finalized
+    );
```
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:353]: remove

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs [hunk 17]
```diff
@@ -285,15 +247,155 @@ impl Datasource<RawEthData> for EthSyncDatasource {
...
+            provider.get_transaction(log.tx_hash),
+            Duration::from_secs(30000)
+        ) else {
+            panic!("Unable to get transaction from provider");
+        };
+
+        data.push((log, block, transaction, is_finalized));
+    }
+
+    println!("Sending {} logs to data sender", data.len());
```
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:381]: remove

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs [hunk 18]
```diff
@@ -304,6 +406,8 @@ impl<E: EthEvent> DataMapper<(E, Block<H256>, Transaction), ProcessedTxnData> fo
         let timestamp_ms = block.timestamp.as_u64() * 1000;
         let gas = transaction.gas;
 
+        println!("Mapping eth event: {:?}", bridge_event);
```
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:409]: remove

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs [hunk 19]
```diff
@@ -285,15 +247,155 @@ impl Datasource<RawEthData> for EthSyncDatasource {
...
+                tx_hash: log.tx_hash,
+                log: log.log,
+            })
+            .collect();
+
+        println!(
+            "Received {} live finalized logs at block: {}",
+            raw_logs.clone().len(),
+            block
+        );
```
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:284]: please remove all print(ln)! in PRs

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs [hunk 20]
```diff
@@ -285,15 +247,155 @@ impl Datasource<RawEthData> for EthSyncDatasource {
...
+
+        process_logs(raw_logs, provider.clone(), data_sender.clone(), block, true)
+            .await
+            .expect("Failed to process logs");
+    }
+
+    panic!("Eth syncer stopped unexpectedly");
```
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:291]:
```suggestion
    panic!("Eth finalized syncer live task stopped unexpectedly");
```

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs [hunk 21]
```diff
@@ -32,27 +32,28 @@ use crate::{
     BridgeDataSource, ProcessedTxnData, TokenTransfer, TokenTransferData, TokenTransferStatus,
 };
 
-type RawEthData = (RawEthLog, Block<H256>, Transaction);
+type RawEthData = (RawEthLog, Block<H256>, Transaction, bool);
```
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:35]: make this a struct so the `bool` is more readable

**Review by longbowlu — COMMENTED** [review]
Could you elaborate test the new insert query?

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs (hunk 12 above)
Comment (Bridgerz) [reply crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:288]: Yes. My idea was to not have `retrieve_and_process_live_finalized_logs` return a result as the task should never die.

**Review by longbowlu — APPROVED** [review]
